src/main/java/edu/harvard/iq/dataverse/harvest/server/xoai/DataverseXoaiItemRepository.java

```diff
@@ -0,0 +1,329 @@
+import io.gdcc.xoai.dataprovider.repository.ResultsPage;
+import io.gdcc.xoai.model.oaipmh.ResumptionToken;
+import io.gdcc.xoai.model.oaipmh.results.record.Metadata;
+import io.gdcc.xoai.xml.EchoElement;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.ArrayList;
+import java.time.Instant;
+import java.util.List;
+import java.util.logging.Logger;
+
+/**
+ *
+ * @author Leonid Andreev
+ * Implements an XOAI "Item Repository". Retrieves Dataverse "OAIRecords" 
+ * representing harvestable local datasets and translates them into 
+ * XOAI "items". 
+ */
+
+public class DataverseXoaiItemRepository implements ItemRepository {
+    private static Logger logger = Logger.getLogger("edu.harvard.iq.dataverse.harvest.server.xoai.DataverseXoaiItemRepository");
+    
+    private OAIRecordServiceBean recordService;
+    private DatasetServiceBean datasetService;
+    private String serverUrl; 
+
+    public DataverseXoaiItemRepository (OAIRecordServiceBean recordService, DatasetServiceBean datasetService, String serverUrl) {
+        super();
+        this.recordService = recordService;
+        this.datasetService = datasetService;
+        this.serverUrl = serverUrl; 
+    }
+    
+    private List<DataverseXoaiItem> list = new ArrayList<DataverseXoaiItem>();
+
+
+    @Override
+    public ItemIdentifier getItem(String identifier) throws IdDoesNotExistException {
+        // This method is called when ListMetadataFormats request specifies 
+        // the identifier, requesting the formats available for this specific record.
+        // In our case, under the current implementation, we need to simply look 
+        // up if the record exists; if it does, all the OAI formats that we serve
+        // should be available for this record. 
+        
+        List<OAIRecord> oaiRecords = recordService.findOaiRecordsByGlobalId(identifier);
+        if (oaiRecords != null && !oaiRecords.isEmpty()) {
+            DataverseXoaiItem xoaiItem = null; 
+            for (OAIRecord oaiRecord : oaiRecords) {
+                // We can return the first *active* record we find for this identifier. 
+                if (!oaiRecord.isRemoved()) {
+                    xoaiItem = new DataverseXoaiItem(oaiRecord);
+                    return xoaiItem; 
+                }
+            }
+        }
+        
+        throw new IdDoesNotExistException();
+    }
+    
+    @Override
+    public Item getItem(String identifier, MetadataFormat metadataFormat) throws HandlerException {
+        logger.fine("getItem; calling findOaiRecordsByGlobalId, identifier " + identifier);
+        
+        if (metadataFormat == null) {
+            throw new NoMetadataFormatsException("Metadata Format is Required");
+        }
+        
+        List<OAIRecord> oaiRecords = recordService.findOaiRecordsByGlobalId(identifier);
+        if (oaiRecords != null && !oaiRecords.isEmpty()) {
+            DataverseXoaiItem xoaiItem = null; 
```

Comment:
This could be a matter of taste. I feel like the way this is implemented here makes sense. This loop logic is not only about waiting to create the item until we know that there's at least one record available; it is also to add the extra sets to the item that was created and populated with metadata when the first record was encountered, when there are multiple records.
(there are multiple records in the database per identifier, one record for every set in which it occurs)